### ISSUE TITLE
[turborepo] turbo prune in v1.7.1 generates turbo.json file that breaks commands

### ISSUE THREAD

**Issue by jplwood** [vercel/turbo#3570]
### What version of Turborepo are you using?

1.7.1

### What package manager are you using / does the bug impact?

npm

### What operating system are you using?

Linux

### Describe the Bug

_originally posted as a comment here: https://github.com/vercel/turbo/issues/3564, but I think it’s actually a separate issue than the one described there_

 My CI dockerfile build randomly started failing today when seemingly nothing changed. I noticed that my dockerfile blindly globally installs the latest turbo in order to prune the repo, so we started using 1.7.1 today. I get the following error on my turbo build step, after I've already pruned my repo:

```
Step 17/33 : RUN npm run build -- --filter=api... &&     npm prune --omit=dev
 ---> Running in 1e638563d2dc

> traveler-recruitment@0.0.0 build
> turbo run build --filter=api...

 ERROR  run failed: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string
Turbo error: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string
The command '/bin/sh -c npm run build -- --filter=api... &&     npm prune --omit=dev' returned a non-zero code: 1
```

Upon trying out the turbo prune locally on `1.7.0` and `1.7.1`, the generated `turbo.json` files are significantly different. Nothing is _wrong_ per se with the version produced by 1.7.1, but  `turbo build` is clearly unhappy about this differently formatted turbo.json file.

<details><summary>1.7.0 turbo.json</summary>

```json
{
  "$schema": "https://turbo.build/schema.json",
  "pipeline": {
    "dev": {
      "dependsOn": ["^build"],
      "outputs": [],
      "cache": false,
      "persistent": true
    },
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["build/**", "dist/**"],
      "outputMode": "new-only"
    },
    "lint": {
      "dependsOn": ["^build"],
      "outputs": [],
      "outputMode": "new-only"
    },
    "test": {
      "outputs": ["coverage/**"],
      "dependsOn": ["^build"],
      "inputs": ["src/**/*.tsx", "src/**/*.ts", "jest.config.ts", "jest-setup.ts"],
      "outputMode": "new-only"
    },
    "test:watch": {
      "cache": false,
      "persistent": true
    },
    "clean": {
      "cache": false
    }
  }
}
```

</details>

<details>
<summary>1.7.1 turbo.json</summary>

```json
{
  "pipeline": {
    "build": {
      "Outputs": { "Inclusions": ["build/**", "dist/**"], "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 3,
      "Persistent": false
    },
    "clean": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": [],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": false
    },
    "dev": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": true
    },
    "lint": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 3,
      "Persistent": false
    },
    "test": {
      "Outputs": { "Inclusions": ["coverage/**"], "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": ["src/**/*.tsx", "src/**/*.ts", "jest.config.ts", "jest-setup.ts"],
      "OutputMode": 3,
      "Persistent": false
    },
    "test:watch": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": [],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": true
    }
  },
  "remoteCache": {}
}
```
</details>

My workaround for now is pinning the turbo version to 1.7.0 for my docker build

### Expected Behavior

`turbo prune` should generate a turbo.json file that can be used by turbo CLI commands

### To Reproduce

run turbo prune on 1.7.1 and 1.7.0, the turbo.json file generated in the `/out/` dir in 1.7.1 doesn’t work with the turbo commands you configured in your root turbo.json. Should be able to confirm this with a docker build in the `with-docker` example, though I haven’t tried that one yet (also it uses yarn, so not positive if this is only NPM)

_I’ll try to get more detail in here tomorrow_

**Comment by eden** [comment]
Seems related to this PR https://github.com/vercel/turbo/pull/3216

**Comment by abenjaminov** [comment]
> Seems related to this PR #3216

in what way exactly is it related?

**Comment by xpluscal** [comment]
Confirm, having the same issue with builds at the moment. Any ideas?

**Comment by sebaplaza** [comment]
Confirmed, i'm having the same issue with `1.7.1`.

Everything is fine going back to `1.7.0`

**Comment by moshest** [comment]
As @sebaplaza mention, I manage to fix this by patching my docker file:

```Dockerfile
# FIXME lock version to v1.7.0: https://github.com/vercel/turbo/issues/3570
RUN yarn global add turbo@1.7.0 
```

Note that if you're using the [official example](https://turbo.build/repo/docs/handbook/deploying-with-docker#example), you only need that on the `AS pruned` image (the one that run the `turbo prune` command).

**Comment by beefancohen** [comment]
somewhat related, we are seeing issues with the pnpm lockfile after pruning our monorepo:

```
#23 [builder 11/15] RUN pnpm fetch
#23 sha256:1729c6ae2ce9354de6b7cf008dd9a417cc0152362eb3f16f2f4fc9e3292ad8c4
#23 0.990 Importing packages to virtual store
#23 1.101  WARN  Broken lockfile: no entry for '/uuid/8.3.2' in pnpm-lock.yaml
#23 1.102  ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  The lockfile is broken! Resolution step will be performed to fix it.
```

**Comment by jplwood** [comment]
@chris-olszewski Just saw you closed this. I just tried installing 1.7.2 in my docker container to try it out and I'm still getting the same error

```
 => [prune_repo 5/7] RUN npm install --global turbo@^1.7.2                                                                                                                                                                       3.6s 
    ...
 => ERROR [build_image 9/9] RUN npm run build -- --filter=api... &&     npm prune --omit=dev                                                                                                                                     1.4s 
------
 > [build_image 9/9] RUN npm run build -- --filter=api... &&     npm prune --omit=dev:
#22 1.247
#22 1.247 > traveler-recruitment@0.0.0 build
#22 1.247 > turbo run build --filter=api...
#22 1.247
#22 1.269 Turbo error: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string
```

And the turbo.json still looks wrong:

<details><summary><strong>1.7.2</strong> <code>turbo.json</code></summary>

```json
{
  "pipeline": {
    "build": {
      "Outputs": { "Inclusions": ["build/**", "dist/**"], "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 3,
      "Persistent": false
    },
    "clean": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": [],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": false
    },
    "dev": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": true
    },
    "lint": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 3,
      "Persistent": false
    },
    "test": {
      "Outputs": { "Inclusions": ["coverage/**"], "Exclusions": null },
      "ShouldCache": true,
      "EnvVarDependencies": [],
      "TopologicalDependencies": ["build"],
      "TaskDependencies": [],
      "Inputs": ["src/**/*.tsx", "src/**/*.ts", "jest.config.ts", "jest-setup.ts"],
      "OutputMode": 3,
      "Persistent": false
    },
    "test:watch": {
      "Outputs": { "Inclusions": null, "Exclusions": null },
      "ShouldCache": false,
      "EnvVarDependencies": [],
      "TopologicalDependencies": [],
      "TaskDependencies": [],
      "Inputs": null,
      "OutputMode": 0,
      "Persistent": true
    }
  },
  "remoteCache": {}
}

```
</details>

Might have been some other problem, unless I'm missing something?

**Comment by joshunrau** [comment]
I am having the same problem as @jplwood using v1.7.2 with Docker. No issue using the same version locally on Mac.

`shell
root@3ec9caa4a587:/app# yarn turbo run build --filter=client
yarn run v1.22.19
$ /app/node_modules/.bin/turbo run build --filter=client
 ERROR  run failed: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string
Turbo error: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`

**turbo.json**
`json
{
  "globalDependencies":[
    ".env"
  ],
  "pipeline":{
    "build":{
      "Outputs":{
        "Inclusions":[
          "dist/**"
        ],
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        "build"
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    },
    "check-types":{
      "Outputs":{
        "Inclusions":null,
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    },
    "dev":{
      "Outputs":{
        "Inclusions":null,
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        "build"
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    },
    "format":{
      "Outputs":{
        "Inclusions":null,
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    },
    "lint":{
      "Outputs":{
        "Inclusions":null,
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        "build"
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    },
    "test":{
      "Outputs":{
        "Inclusions":null,
        "Exclusions":null
      },
      "ShouldCache":true,
      "EnvVarDependencies":[
        
      ],
      "TopologicalDependencies":[
        "build"
      ],
      "TaskDependencies":[
        
      ],
      "Inputs":null,
      "OutputMode":0,
      "Persistent":false
    }
  },
  "remoteCache":{
    
  }
}
`

**Comment by LukasCornille** [comment]
I believe this will be fixed in the 1.7.3 release. Viewing the commit history, this was fixed after the 1.7.2 release 😅.

<img width="525" alt="image" src="https://user-images.githubusercontent.com/36887321/216274833-78a8ed9f-9cfc-4cb3-88e4-939690a64ce9.png">

**Comment by xpluscal** [comment]
So until when it’s expected to stick to 1.7.0?

Sent via Superhuman iOS ( ***@***.*** )

**Comment by HQ92** [comment]
I'm getting the same error recently on GitHub Actions with turbo@1.7.0

turbo run deploy --filter=frontend
"ERROR  run failed: turbo.json: json: cannot unmarshal number into Go struct field rawTurboJSON.pipeline of type string"

**Comment by maximveksler** [comment]
As instructed by @chris-olszewski, I'm asking to reopen this issue. More details at https://github.com/vercel/turbo/issues/3645#issuecomment-1420648567

**Comment by jplwood** [comment]
I just tested this again against turbo v1.7.4 and the issue seems to be resolved.

**Comment by chris-olszewski** [comment]
@maximveksler Can you quick check if `1.7.4` fixes the issue? If not can you provide a reproduction repo?

**Comment by maximveksler** [comment]
@chris-olszewski, unfortuntly same effect.

```
(base) m@m-inspiron5567:~/code/allbetter/Core$ turbo --version
1.7.4
(base) m@m-inspiron5567:~/code/allbetter/Core$ turbo prune --scope="better" --docker
Generating pruned monorepo for better in /home/m/code/allbetter/Core/out
 - Added better
 - Added better-config
 - Added generator-app
 - Added middlewares
 - Added utils
 ERROR  failed to read turbo.json: Could not find turbo.json. Follow directions at https://turbo.build/repo/docs to create one: file does not exist
Turbo error: failed to read turbo.json: Could not find turbo.json. Follow directions at https://turbo.build/repo/docs to create one: file does not exist
```

**Comment by moshest** [comment]
@maximveksler I don't think this related. The problem solved for me.

I think your error is because you don't have the `turbo.json` on the right location or it not imported to docker. Check your `.dockerignore` file.

**Comment by maximveksler** [comment]
@moshest I agree, as mentioned here https://github.com/vercel/turbo/issues/3645#issuecomment-1425157546 the point I'm making is that the regression is that up until 1.7.0 turbo with no turbo.json worked fine for the prune use case.

I think this issue should be closed, and #3645 be reopened.
